File: 98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <climits>
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (root == NULL) return true;
        return check(root, LONG_MIN, LONG_MAX);
    }
    
    bool check(TreeNode* root, long long int lower, long long int upper)
    {
        return root->val > lower && root->val < upper && ((root->left ? check(root->left, lower, root->val) : true) && (root->right ? check(root->right, root->val, upper) : true));
    }
};

File: 98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import sys
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        return self.check(root, -sys.maxint, sys.maxint)
        
    def check(self, root, lower, upper):
        return root.val < upper and root.val > lower and (self.check(root.left, lower, root.val) if root.left else True) and (self.check(root.right, root.val, upper) if root.right else True)
